nupic/research/frameworks/vernon/experiments/base_experiment.py

```diff
@@ -0,0 +1,184 @@
+                                  before it
+            - launch_time: time the config was created (via time.time). Used to report
+                           wall clock time until the first batch is done.
+                           Default: time.time() in this setup_experiment().
+        """
+        # Configure logging related stuff
+        log_format = config.get("log_format", logging.BASIC_FORMAT)
+        log_level = getattr(logging, config.get("log_level", "INFO").upper())
+        console = logging.StreamHandler()
+        console.setFormatter(logging.Formatter(log_format))
+        self.logger = logging.getLogger(config.get("name", type(self).__name__))
+        self.logger.setLevel(log_level)
+        self.logger.addHandler(console)
+        self.progress = config.get("progress", False)
+        self.launch_time = config.get("launch_time", time.time())
+        self.logdir = config.get("logdir", None)
+
+        if config.get("disable_logger", False):
+            self.logger.disabled = True
+            self.progress = False
+
+        # Configure seed
+        self.seed = config.get("seed", 42)
+        set_random_seed(self.seed, False)
+
+        self.logger.info("Execution order: %s",
+                         pformat(self.get_execution_order()))
+
+    @abc.abstractmethod
+    def run_iteration(self):
+        """
+        Run one training iteration of the experiment and return some sort of
+        result.
+        """
+
+    @abc.abstractmethod
+    def should_stop(self):
+        """
+        Whether or not the experiment should stop. Usually determined by the
+        number of epochs but customizable to any other stopping criteria
+        """
+
+    def stop_experiment(self):
+        """
+        Perform any needed cleanup.
+        """
+
+    def run_pre_experiment(self):
+        """
+        Evaluate the model before performing any training and return some sort
+        of result. If the experiment is configured to skip the pre_experiment,
+        return None.
+        """
+        return None
+
+    @classmethod
+    def insert_pre_experiment_result(cls, result, pre_experiment_result):
+        """
+        Modify result to incorporate the pre_experiment_result. By default, this
+        method performs no update. (This may be sufficient, for example if the
+        pre_experiment_result has already been printed to the console.)
+
+        :param pre_experiment_results: The return value of pre_experiment
+        """
+
+    def get_state(self):
+        """
+        Get experiment serialized state as a dictionary of  byte arrays
+        :return: dictionary with "model", "optimizer" and "lr_scheduler" states
+        """
+        return {}
+
+    def set_state(self, state):
+        """
+        Restore the experiment from the state returned by `get_state`
+        :param state: dictionary with "model", "optimizer", "lr_scheduler"
+        states
+        """
+
```

Comment:
A maybe controversial opinion: should we create an option for the experiment to train itself, without the Trainable? That would be a very vanilla function (`while not should stop: run epoch; print results`). I think that would make a lot easier for beginners to run simple experiments on notebooks, just instantiate an experiment class and call `exp.train()`.

No changes to the whole Trainable thing, it is just becomes another option. Lots of frameworks have more than one approach of doing something, usually a simpler version and a more complex one for power users (Keras for example).